src/ace/managers/viewport/tilebuffer.c

```diff
@@ -658,24 +682,60 @@ void tileBufferRedrawAll(tTileBufferManager *pManager) {
 	PLANEPTR pDstPlane = pManager->pScroll->pBack->Planes[0];
 	tTileBufferTileIndex **pTileData = pManager->pTileData;
 	UWORD uwBltsize = tileBufferSetupTileDraw(pManager);
+#if defined ACE_DEBUG
+	if ((uwBltsize & BLIT_WORDS_NON_INTERLEAVED_BIT) && !ubNonInterleavedBlit) {
+		logWrite("ERR: you called tileBufferRedrawAll%s, but the buffers are %s",
+			ubNonInterleavedBlit ? "NonInterleaved" : "Interleaved",
+			ubNonInterleavedBlit ? "interleaved" : "not interleaved"
+		);
+	}
+#endif
 	UWORD uwTileOffsX = (wStartX << ubTileShift);
-	UWORD uwDstOffsStep = ubTileSize / 8;
+	UWORD uwMarginedHeight = pManager->uwMarginedHeight;
+	UWORD uwWrapAroundY = MIN(uwEndY, uwMarginedHeight >> ubTileShift);
+	UWORD uwBfrY = wStartY << ubTileShift;
 	systemSetDmaBit(DMAB_BLITHOG, 1);
-	for (UWORD uwTileY = wStartY; uwTileY < uwEndY; ++uwTileY) {
-		UWORD uwTileCurr = wStartX;
-		ULONG ulDstOffs = uwDstBytesPerRow * uwTileOffsY + uwTileOffsX / 8;
-		while(uwTileCurr < uwEndX) {
+	Disable(); // Disable and clear CPU caches (if any)
```

Comment:
I'm not sure if `Disable()`/`Enable()` is proper way to handle it, especially that it also touches multitasking stuff according to docs and we might have OS dead on our end.

This is tricky, because I'd see it as `systemDisableCache()` that would do `Disable()` when OS is on, and some inline assembly when OS is off. Or, we could assume that ownblit is good enouhh with OS on and just use the inline asm to disable caches. We could also make a debug check for OS being turned off as a prerequisite for hot path.

In the past, I had some stability issues when using Forbid/Permit or Disable/Enable but that was a long ago and I don't remember details. Docs say they shouldn't be used for too long, and I'm not sure how long is too long.

Reply:
This one I still need to think about. Also, `CacheClearU` seems to be a KS2 function, so that's also a problem.